src/expr/src/relation/canonicalize.rs

```diff
@@ -83,6 +87,123 @@ pub fn canonicalize_predicates(predicates: &mut Vec<MirScalarExpr>, input_type:
             predicates.push(expr);
         }
     }
+
+    // 3) Reduce across `predicates`.
+    // If a predicate `p` cannot be null, and `f(p)` is of type bool
+    // (`f(p)` is allowed to be nullable), then the predicate `p & f(p)` is
+    // equal to `p & f(true)`, and `!p & f(p)` is equal to `!p & f(false)`.
+    // For any index i, the `Vec` of predicates `[p1, ... pi, ... pn]` is
+    // equivalent to the single predicate `pi & (p1 & ... & p(i-1) & p(i+1)
+    // ... & pn)`.
+    // Thus, if `pi` (resp. `!pi`) cannot be null, it is valid to replace with
+    // `true` (resp. `false`) every subexpression in
+    // `(p1 & ... & p(i-1) & p(i+1) ... & pn)` that is equal to `pi`.
+
+    // (As for the reduction does not work if `p` can be null: if `p` is null,
+    // then p & f(p) becomes `null & f(null)`. `null & f(null)` is not
+    // necessarily equal to `null & f(true)` because `null & true = null`
+    // and `null & false = false`.)
```

Comment:
> (we are in the special circumstance of evaluating predicates, rather than just performing general expression simplifications)

That is a good point. I was hesitant about collapsing the `null` and `false` answers, but now that I think about it, we don't have any transforms that would propagate information in such a way that Filter `p&q` turns into something like Filter `!(p&q)`.

Reply:
I could totally be wrong, too! :D